### ISSUE TITLE
Request: ability to change Avahi-advertised hostname/IP

### ISSUE THREAD

**Issue by airdrummingfool** [mbentley/docker-timemachine#99]
**Is your feature request related to a problem? Please describe.**
I am using this container on a host that already runs Avahi. I did not set up a macvlan or use host networking, but because the host `avahi-daemon` has `enable-reflector=yes` in the config, it is reflecting advertisements from the Docker network to my wider LAN, which means my networked Macs can see the container in their Time Machine settings (this is good!). However, because the container is advertising itself on the network it knows (Docker's network), the hostname is specific to the Docker network, and not accessible outside of the host computer. E.g. the hostname in the advertisement is `abcd1234.local`, and the IP address it maps to is `172.18.0.9`, while the actual host IP on my local network is `192.168.1.10`.

**Describe the solution you'd like**
It would be great if an environment variable was available to configure an override for the hostname that is advertised. For example, if I could set `ADVERTISED_HOST=MyServer.local`, which would then advertise the TimeMachine service as available at `MyServer.local` instead of `abcd1234.local`.

**Additional context**
To be clear, I'm not sure it's possible to override an advertisement in this way. I have done something similar in other contexts, including setting a different published server URL for Jellyfin, and setting a different advertised host for Home Assistant when pairing to a HomeKit accessory.

If this feature request does work, it will provide a simple way to avoid Avahi conflicts with the host, without having to configure a Docker `macvlan`.

**Comment by airdrummingfool** [comment]
Small update: this appears to be feasible by editing `/etc/avahi/services/smbd.service` and replacing the `%h` in the `<name>` field.

So `<name replace-wildcards="yes">%h</name>` becomes `<name replace-wildcards="yes">MyServer</name>`

I will test this and then work up a PR!

**Comment by mbentley** [comment]
Hmm, doing Avahi reflecting could be interesting.  I know nothing about it but if there is anything on an example configuration, I can always work on a proof of concept for it.

**Comment by airdrummingfool** [comment]
I got part-way there: replacing the `<name>` entry in the `smbd.service` file changed the hostname that was advertised, but that entry just points to an `SRV` entry that has the actual container hostname in it. Here are the entries:

```
_smb._tcp                               PTR     80a176e5917f._smb._tcp
80a176e5917f._smb._tcp                  SRV     0 0 445 80a176e5917f.local. ; Replace with unicast FQDN of target host
80a176e5917f._smb._tcp                  TXT     ""
```
The `_smb._tcp` `PTR` records is the one that is easy to override.

I was not able to figure out how to change the hostname of the `SRV` records to an existing host. The advertised hostname can be overridden in `/etc/avahi/avahi-daemon.conf` (use the `host-name` parameter in the `[server]` section), but it seemed to sense the collision with an existing host on the network and added a numerical suffix to the hostname: `MyServer` became `MyServer (2)`. I couldn't figure out how to override that `SRV` entry to use an existing hostname.

The next thing I did was arguably simpler: I just pasted the entire `smbd.service` file onto my Docker host, which is already running Avahi. It picked it up and started advertising the `smb` service with the correct hostname (since it's running on the host itself) and a client was able to find the Time Machine server and connect.

So, this is the solution I will be going with for now - it seems simpler than using the reflector (which sometimes causes other network weirdness).

Is there an easy way to completely disable avahi in the container? I don't need it anymore since I am advertising on the host directly, but I need to keep my reflector on because there are other containers that advertise.

**Comment by airdrummingfool** [comment]
OK, I think I figured it out. This **can** be done with just a few modifications to the `smbd.service` XML file.

1. Change the entry in the `<name>` field to be the desired `hostname`
2. Add a `<host-name>` entry to each service with the desired `hostname.domain`

e.g.
```xml
<?xml version="1.0" standalone='no'?><!--*-nxml-*-->
<!DOCTYPE service-group SYSTEM "avahi-service.dtd">

<service-group>
  <name replace-wildcards="yes">MyServer</name>
  <service>
    <type>_smb._tcp</type>
    <host-name>MyServer.local</host-name>
    <port>445</port>
  </service>
  <service>
    <type>_device-info._tcp</type>
    <host-name>MyServer.local</host-name>
    <port>9</port>
  <txt-record>model=TimeCapsule8,119</txt-record>
  </service>
  <service>
    <type>_adisk._tcp</type>
    <host-name>MyServer.local</host-name>
    <port>9</port>
    <txt-record>sys=adVF=0x100</txt-record>
    <txt-record>dk0=adVN=TimeMachine,adVF=0x82</txt-record>
  </service>
</service-group>
```

Note that the TLD (`.local`) should not be added to the `<name>` field, but it must be included in the `<host-name>` fields.

Documentation for `<host-name>`:
> The host name of the host that provides this service. This should be a host that is resolvable by multicast or unicast DNS. Please note that you need to specify a fully-qualified domain name (FQDN) here, i.e. .local is not appended implicitly!

Here is the output of `dns-sd -Z _smb` after the above changes:
```
_smb._tcp                                       PTR     MyServer._smb._tcp
MyServer._smb._tcp                             SRV     0 0 445 MyServer.local.
MyServer._smb._tcp                             TXT     ""
```
This does not cause a collision with existing hostnames the way that editing `/etc/avahi/avahi-daemon.conf` does.

Assuming my testing goes well, is this something you'd be interested in a PR for? I think it would offer a simpler way of integrating into an environment with other containers that used mDNS. That being said, I still think the option to fully disable avahi in the container would be nice to have as well.

**Comment by airdrummingfool** [comment]
Fixed in #100  😁
